### PR TITLE
Uploads validation rules

### PR DESCRIPTION
**Pull request by pxpm**
We have now two custom rules to validate upload fields.. `ValidUpload` and `ValidUploadMultiple`. 

```php
// removed, outdated examples.
```

~Both accept the `arrayRules()` and `fileRules()`.~

~Like the name says, the `fileRules()` are applied on the uploaded files while the `arrayRules` are applied on the "group of uploaded files".~

~Write your `arrayRules` and `fileRules` exactly as you would expect in Laravel, except the need to have different rules for "create" and "update" (Looking at you `sometimes`).~

Read docs: https://github.com/Laravel-Backpack/docs/pull/422

Enjoy. 😎

### CODE REVIEW

File: src/app/Library/CrudPanel/Traits/Validation.php [hunk 1]
```diff
@@ -367,6 +250,77 @@ private function setupFieldValidation($field, $parent = false)
...
+        return $messages;
+    }
+
+    /**
+     * Return the rules for the fields and subfields in the current crud panel.
+     *
+     * @param  array  $fields
+     * @return array
+     */
+    private function getValidationRulesFromFieldsAndSubfields($fields)
```
Comment (tabacitu) [inline src/app/Library/CrudPanel/Traits/Validation.php:295]: Note to self: this hasn't changed at all, just moved in a different place in the same file.

File: src/app/Library/CrudPanel/Traits/Validation.php [hunk 2]
```diff
@@ -367,6 +250,77 @@ private function setupFieldValidation($field, $parent = false)
         }
     }
 
+    /**
+     * Return the messages for the fields and subfields in the current crud panel.
+     *
+     * @param  array  $fields
+     * @return array
+     */
+    private function getValidationMessagesFromFieldsAndSubfields($fields)
```
Comment (tabacitu) [inline src/app/Library/CrudPanel/Traits/Validation.php:259]: Note to self: this hasn't changed at all, just moved in a different place in the same file.

File: src/app/Library/CrudPanel/Traits/Validation.php [hunk 3]
```diff
@@ -397,4 +351,83 @@ private function getValidationRulesAndMessagesFromField($field, $parent = false)
...
+
+    /**
+     * Return an array containing the request rules and the field/controller rules merged.
+     * The rules in request will take precedence over the ones in controller/fields.
+     *
+     * @param  \Illuminate\Http\Request  $request
+     * @param  array  $rules
+     * @return array
+     */
+    private function mergeRules($request, $rules)
```
Comment (tabacitu) [inline src/app/Library/CrudPanel/Traits/Validation.php:363]: Note to self: this hasn't changed at all, just moved in a different place in the same file.

File: src/app/Library/CrudPanel/Traits/Validation.php [hunk 4]
```diff
@@ -397,4 +351,83 @@ private function getValidationRulesAndMessagesFromField($field, $parent = false)
...
+        return array_merge($rules, $extendedRules);
+    }
+
+    /**
+     * Return the request rules as an array of rules if developer provided a rule string configuration.
+     *
+     * @param  \Illuminate\Http\Request  $request
+     * @return array
+     */
+    private function getRequestRulesAsArray($request)
```
Comment (tabacitu) [inline src/app/Library/CrudPanel/Traits/Validation.php:385]: Note to self: this hasn't changed at all, just moved in a different place in the same file.

File: src/app/Library/CrudPanel/Traits/Validation.php [hunk 5]
```diff
@@ -397,4 +351,83 @@ private function getValidationRulesAndMessagesFromField($field, $parent = false)
...
+
+    /**
+     * Checks if the request is valid against the rules.
+     *
+     * @param  array  $rules
+     * @param  array  $messages
+     * @param  \Illuminate\Http\Request|null  $request
+     * @return \Illuminate\Http\Request
+     */
+    private function checkRequestValidity($rules, $messages, $request = null)
```
Comment (tabacitu) [inline src/app/Library/CrudPanel/Traits/Validation.php:403]: Note to self: this hasn't changed at all, just moved in a different place in the same file.

File: src/app/Library/CrudPanel/Traits/Validation.php [hunk 6]
```diff
@@ -306,16 +180,25 @@ public function setRequiredFields($classOrRulesArray)
...
-                ) {
-                    if (Str::contains($key, '.')) {
-                        $key = Str::dotsToSquareBrackets($key, ['*']);
-                    }
+            foreach ($rules as $key => $validationRules) {
+                if (is_string($validationRules)) {
+                    $validationRules = explode('|', $validationRules);
+                }
+                foreach ($validationRules as $rule) {
+                    if (is_a($rule, ValidBackpackUpload::class, true)) {
```
Comment (tabacitu) [inline src/app/Library/CrudPanel/Traits/Validation.php:188]: From what I can tell... this is tying a core function (`public function setRequiredFields()`) to a custom validator we've just created - `ValidBackpackUpload` 👀

red-flag-stop-gif-by-richard-childress-racing

Could you please explain why this is needed @pxpm ? I guess it's related to what you tried to explain to me on Discord, but I didn't understand there either 😀

1) ELI5 why this is needed please. 
2) Why do we consider `ValidBackpackUpload` is the same thing as `required` - to show the asterisk? 
3) Why don't people have to do both `required` and `ValidBackpackUpload` when that is the case? Why does `ValidBackpackUpload` also mean required? What about when you want it to be a valid upload, but only if present? Like if you would combine `nullable` and `ValidBackpackUpload` - in my mind?

File: src/app/Library/Uploaders/Validation/ValidBackpackUpload.php [hunk 7]
```diff
@@ -0,0 +1,160 @@
+<?php
```
Comment (tabacitu) [inline src/app/Library/Uploaders/Validation/ValidBackpackUpload.php:1]: This class is SO FREAKING CLEAN.

slow_clap_cap

File: src/app/Library/Uploaders/Validation/ValidBackpackUpload.php [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+     * Run the validation rule.
+     *
+     * @param  string  $attribute
+     * @param  mixed  $value
+     * @param  Closure(string): \Illuminate\Translation\PotentiallyTranslatedString  $fail
+     * @return void
+     */
+    public function validate(string $attribute, mixed $value, Closure $fail): void
+    {
+    }
```
Comment (tabacitu) [inline src/app/Library/Uploaders/Validation/ValidBackpackUpload.php:48]: Why does this method exist but is empty? 👀

File: src/app/Library/Uploaders/Validation/ValidUploadMultiple.php [hunk 9]
```diff
@@ -0,0 +1,66 @@
+<?php
```
Comment (tabacitu) [inline src/app/Library/Uploaders/Validation/ValidUploadMultiple.php:1]: happy-the-office-gif-by-cameo

File: src/app/Library/Uploaders/Validation/ValidBackpackUpload.php (hunk 8 above)
Comment (pxpm) [reply src/app/Library/Uploaders/Validation/ValidBackpackUpload.php:48]: Just so that I knew the signature when working 🙃 I will delete it, it's the overwrite from the parent class.

File: src/app/Library/CrudPanel/Traits/Validation.php (hunk 6 above)
Comment (pxpm) [reply src/app/Library/CrudPanel/Traits/Validation.php:188]: Hey @tabacitu 

`ValidBackpackUpload` is the "validation rule" that developers define. There "real rules" are inside `->arrayRules()` when the rule is a `ValidBackpackUpload` rule (both `ValidUploadMultiple` and `ValidDropzone` are extensions of the `ValidBackpackUpload` rule).

So what I am doing here, is checking if the rule is a `ValidBackpackUpload`, in case it is, we check the "requiredeness" 😎 inside the rules in `arrayRules()`. 

```php
'attachments' => ['required', ValidUploadMultiple::make()->arrayRules('min:2|max:3')->fileRules(...)]
'attachments' => [ValidUploadMultiple::make()->arrayRules('required|min:2|max:3')->fileRules(...)] // the same as above

// answering your question: I want it to be a valid upload, but only if present.
'attachments' => [ValidUploadMultiple::make()->fileRules(...)],
//or when present min 1 file
'attachments' => [ValidUploadMultiple::make()->arrayRules('min:1|max:3')->fileRules(...)],
'attachments' => ['min:1', 'max:3', ValidUploadMultiple::make()->fileRules(...)], // the same as above
``` 

In that foreach we are iterating the `attachments` validation rules, so for the example: 
`'attachments' => ['required', ValidUploadMultiple::make()->arrayRules('min:2|max:3')->fileRules(...)]`

- the first iteration would rule `required` rule and follow the previous check.
- the second and last would be `ValidUploadMultiple` rule, in that case we will check if the `required` is inside the `arrayRules`

This does not automatically set required because it's a `ValidBackpackUpload`. It handles them and set required if they are required from the array rules  👍

Comment (pxpm) [reply src/app/Library/CrudPanel/Traits/Validation.php:188]: I am not sure I've followed you here. 

Are you suggesting something like renaming `ValidBackpackUpload`  to `ValidArrayData` ? So that we can have:
```php
class ValidUploadMultiple extends ValidArrayData {}
class ValidRepeatable extends ValidArrayData {}
```
If that's the case I agree with you I guess. But ValidArrayData should not be the "rule", only the `ValidUploadMultiple` etc are rules correct?
Like:
```php
class ValidArrayData { }
class ValidUploadMultiple extends ValidArrayData implements ValidationRule, DataAwareRule, ValidatorAwareRule {}
```
Cheers

Comment (pxpm) [reply src/app/Library/CrudPanel/Traits/Validation.php:188]: I think if we do this the `ValidUploadMultiple` class would need to have `getData()`, `getValidator()` and the "custom validator methods" that are used in other validators like `ValidDropzone` and that would bloat up the files so much. 

So I think we should still keep the `ValidArrayData` as the "validation rule", and the others extend it. 

```php
class ValidArrayData implements ValidationRule, DataAwareRule, ValidatorAwareRule {}
class ValidUploadMultiple extends ValidArrayData 
```

Comment (tabacitu) [reply src/app/Library/CrudPanel/Traits/Validation.php:188]: Yup - seems like we're on the same page. The only bad feeling I have is about the name - maybe we call it directly `ValidArray` instead of `ValidArrayData`? "data" doesn't really say anything, does it?

File: src/app/Library/Validation/Rules/ValidArray.php [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+    {
+        if (! is_array($value)) {
+            try {
+                $value = json_decode($value, true);
+            } catch (\Exception $e) {
+                $fail('Unable to determine the value type.');
+
+                return;
+            }
+        }
```
[the remark this answers is not in the document]
Comment (tabacitu) [reply src/app/Library/Validation/Rules/ValidArray.php:34]: Maaaaybe. I don't know

File: src/app/Library/Validation/Rules/BackpackCustomRule.php [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+abstract class BackpackCustomRule implements ValidationRule, DataAwareRule, ValidatorAwareRule
+{
+    /**
+     * @var \Illuminate\Contracts\Validation\Validator
+     */
+    protected $validator;
+
+    protected array $data;
+
+    public array $attributeRules = [];
```
[the remark this answers is not in the document]
Comment (tabacitu) [reply src/app/Library/Validation/Rules/BackpackCustomRule.php:26]: What do you think about this @pxpm ?

[the remark this answers is not in the document]
Comment (tabacitu) [reply src/app/Library/Validation/Rules/BackpackCustomRule.php:26]: Or maybe it'd make sense to keep `fieldRules()` , `fileRules()` and `itemRules()` as part of the API, but behind the scenes they'd work with different names?
- `attributeRules`
- `nestedRules`

I'm having a difficult time wrapping my head around a good name structure for this... Any proposal that would fix this @pxpm ?

**Review by tabacitu — CHANGES_REQUESTED** [review]

File: src/app/Library/Validation/Rules/ValidUpload.php [hunk 12]
```diff
@@ -0,0 +1,47 @@
+<?php
+
+namespace Backpack\CRUD\app\Library\Validation\Rules;
+
+use Closure;
+use Illuminate\Support\Facades\Validator;
+
+class ValidUpload extends BackpackCustomRule
+{
+    public array $fileRules = [];
```
[the remark this answers is not in the document]
Comment (pxpm) [reply src/app/Library/Validation/Rules/ValidUpload.php:10]: Same as above!

File: src/app/Library/Validation/Rules/ValidArray.php [hunk 13]
```diff
@@ -0,0 +1,126 @@
...
+use Closure;
+use Illuminate\Contracts\Validation\Rule;
+use Illuminate\Contracts\Validation\ValidationRule;
+use Illuminate\Support\Facades\Validator;
+
+abstract class ValidArray extends BackpackCustomRule
+{
+    public array $itemRules = [];
+
+    public array $namedItemRules = [];
```
[the remark this answers is not in the document]
Comment (pxpm) [reply src/app/Library/Validation/Rules/ValidArray.php:14]:
```php
// itemRules
'array.*' => [general array rules]

// namedItemRules
'array.name' => [one rule],
'array.description' => 'other rule'
```

I guess it stood to you because they are only used in ValidNestedArray, I will move them there. 👍

File: src/app/Library/Validation/Rules/BackpackCustomRule.php [hunk 14]
```diff
@@ -0,0 +1,121 @@
...
+        ], $this->validator->customMessages, $this->validator->customAttributes);
+
+        if ($validator->fails()) {
+            foreach ($validator->errors()->messages()[$attribute] as $message) {
+                $fail($message)->translate();
+            }
+        }
+    }
+
+    protected static function prepareRules($rules)
```
Comment (tabacitu) [inline src/app/Library/Validation/Rules/BackpackCustomRule.php:109]: Isn't this actually `getRulesArray()`? That's what it does, right? Whatever the rules are, it returns an array of rules. Right? This stood out to me because with a name like `prepareRules()` I'd expect it to be a procedure (return nothing) not a function (return something).

```suggestion
    protected static function getRulesArray($rules)
```

Comment (tabacitu) [reply src/app/Library/Validation/Rules/BackpackCustomRule.php:109]: Alternatively, if we don't want to confuse this with a GETTER (since it's a static method), we could call it `convertRulesToArray()` or `turnRulesIntoArray()` something like that. Right?

**Review by tabacitu — CHANGES_REQUESTED** [review]

**Comment by tabacitu** [comment]
Tested both rules, they work really well! 🎉 Time. To. MERGE! Still a few small improvements that should be done to docs and Uploaders, noted in https://github.com/Laravel-Backpack/docs/pull/421/files . But they don't really affect this branch, so I'm going to merge 🎉

Thanks Pedro!
